<h1>Descriptive Statistics of Variables other than "Text" and "Entities"</h1>

<h3>Packages & Options</h3>

In [1]:
import pandas as pd
pd.set_option('display.max_colwidth', -1)

<h3>Import & Prepare Data for Analysis</h3>

In [30]:
df = pd.read_csv('20190313_TW_MigrantMigration.csv')
df[1:2]

id_str      from_user  \
1  1.102720e+18  BlueelvisEric   

                                                                                                                                                                                                  text  \
1  RT @1776Stonewall: https://t.co/CyucESMlFM  Texas health officials report that nearly 200 people contracted mumps in migrant detention facilities located across the state so far this fiscal year.   

                       created_at                 time geo_coordinates  \
1  Mon Mar 04 23:59:40 +0000 2019  04/03/2019 23:59:40  NaN              

  user_lang  in_reply_to_user_id_str in_reply_to_screen_name  \
1  en       NaN                       NaN                      

   from_user_id_str  in_reply_to_status_id_str  \
1  3.109106e+09     NaN                          

                                                                                 source  \
1  <a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>   

                                                             profile_image_url  \
1  http://pbs.twimg.com/profile_images/1062099415421411333/EZCmOCeu_normal.jpg   

   user_followers_count  user_friends_count   user_location  \
1  940                   225.0               top your momma   

                                                      status_url  \
1  http://twitter.com/BlueelvisEric/statuses/1102720294035353606   

                                                                                                                                                                                                                                                                                                                                                                                                                entities_str  
1  {"hashtags":[],"symbols":[],"user_mentions":[{"screen_name":"1776Stonewall","name":"Stonewall Jackson","id":800086706678419500,"id_str":"800086706678419456","indices":[3,17]}],"urls":[{"url":"https://t.co/CyucESMlFM","expanded_url":"https://www.breitbart.com/border/2019/03/04/200-cases-of-mumps-confirmed-in-texas-migrant-detention-centers/","display_url":"breitbart.com/border/2019/03…","indices":[19,42]}]}

In [31]:
#Drop the "text" and "entities" columns that Sunyam is working on. 
#Also dropping time because it is redundant with "created_at"
df = df.drop(['text','time','entities_str'], axis=1)
df[1:2]

id_str      from_user                      created_at  \
1  1.102720e+18  BlueelvisEric  Mon Mar 04 23:59:40 +0000 2019   

  geo_coordinates user_lang  in_reply_to_user_id_str in_reply_to_screen_name  \
1  NaN             en       NaN                       NaN                      

   from_user_id_str  in_reply_to_status_id_str  \
1  3.109106e+09     NaN                          

                                                                                 source  \
1  <a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>   

                                                             profile_image_url  \
1  http://pbs.twimg.com/profile_images/1062099415421411333/EZCmOCeu_normal.jpg   

   user_followers_count  user_friends_count   user_location  \
1  940                   225.0               top your momma   

                                                      status_url  
1  http://twitter.com/BlueelvisEric/statuses/1102720294035353606

In [32]:
#Rename remaining columns
df = df.rename(columns={"id_str":"tweetID",
                        "from_user":"authorScreen",
                        "created_at":"tweetTimeDate",
                        "geo_coordinates":"tweetGeo",
                        "user_lang":"authorLang",
                        "in_reply_to_user_id_str":"authorID_original",
                        "in_reply_to_screen_name":"authorScreen_original",
                        "from_user_id_str":"authorID",
                        "in_reply_to_status_id_str":"tweetID_original",
                        "source":"tweetMethod", 
                        "profile_image_url":"authorPicture",
                        "user_followers_count":"authorFollower",
                        "user_friends_count":"authorFriend",
                        "user_location":"authorLocation","status_url":"tweetURL"})

df[1:2]

tweetID   authorScreen                   tweetTimeDate tweetGeo  \
1  1.102720e+18  BlueelvisEric  Mon Mar 04 23:59:40 +0000 2019  NaN       

  authorLang  authorID_original authorScreen_original      authorID  \
1  en        NaN                 NaN                   3.109106e+09   

   tweetID_original  \
1 NaN                 

                                                                            tweetMethod  \
1  <a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>   

                                                                 authorPicture  \
1  http://pbs.twimg.com/profile_images/1062099415421411333/EZCmOCeu_normal.jpg   

   authorFollower  authorFriend  authorLocation  \
1  940             225.0         top your momma   

                                                        tweetURL  
1  http://twitter.com/BlueelvisEric/statuses/1102720294035353606

In [33]:
#Group author and tweet variables together in the dataset.
df = df.sort_index(axis=1)
df[1:2]

authorFollower  authorFriend      authorID  authorID_original authorLang  \
1  940             225.0         3.109106e+09 NaN                 en          

   authorLocation  \
1  top your momma   

                                                                 authorPicture  \
1  http://pbs.twimg.com/profile_images/1062099415421411333/EZCmOCeu_normal.jpg   

    authorScreen authorScreen_original tweetGeo       tweetID  \
1  BlueelvisEric  NaN                   NaN      1.102720e+18   

   tweetID_original  \
1 NaN                 

                                                                            tweetMethod  \
1  <a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>   

                    tweetTimeDate  \
1  Mon Mar 04 23:59:40 +0000 2019   

                                                        tweetURL  
1  http://twitter.com/BlueelvisEric/statuses/1102720294035353606

<h3>Data Overview</h3>

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123524 entries, 0 to 123523
Data columns (total 15 columns):
authorFollower           123524 non-null int64
authorFriend             123286 non-null float64
authorID                 123524 non-null float64
authorID_original        9666 non-null float64
authorLang               123524 non-null object
authorLocation           85206 non-null object
authorPicture            123524 non-null object
authorScreen             123524 non-null object
authorScreen_original    9666 non-null object
tweetGeo                 45 non-null object
tweetID                  123524 non-null float64
tweetID_original         9158 non-null float64
tweetMethod              123524 non-null object
tweetTimeDate            123524 non-null object
tweetURL                 123524 non-null object
dtypes: float64(5), int64(1), object(9)
memory usage: 14.1+ MB


<p>The data contain 123,524 observations for 15 variables: nine that describe the author of the tweet and six that describe the tweet. </p>
<p>The author variables are:
<ul>
    <li>author ID and screen name</li>
    <li>author location (from profile)</li>
    <li>author follower count (users following the author)</li>
    <li>author friend count (users the author follows)</li>
    <li>author profile picture (URL)</li>
    <li>screen name and ID of the original author (if tweet is a reply)</li></ul>
The tweet variables are:
<ul>
    <li>tweet ID</li>
    <li>tweet time and date</li>
    <li>tweet URL</li>
    <li>tweet method (e.g., HTML indicating whether author tweeted via web app, phone, Facebook)</li>
    <li>tweet geographical coordinates (if available)</li>
    <li>ID of the original tweet (if tweet is a reply)</li></ul></p>
<p>There is no missing data for authorFollower, authorID, authorLang, authorPicture, authorScreen, tweetID, tweetMethod, tweetTimeDate and tweetURL. There is some missing data for authorFriend, authorID_original, authorScreen_original and tweetID_original. There is a lot of missing data for tweetGeo.</p>
<p>Six of the variables contain numerical data (authorFollower, authorFriend, authorID, authorID_original, tweetID and tweetID_original); however, the last four of these are numerical representations of things (either Twitter accounts or tweets). As such, the data should be treated as object rather than numerical data. The remaining nine variables contain object (string) data. Five out of the six numerical variables are of the type float64, which is a format allowing a fractional number. One is of the type int64, which includes only whole numbers. For all of the variables, the int64 data type makes more sense because the things being represented (Twitter accounts and tweets) do not come in parts.</p>
<p>Of the object-type variables, some might be more appopriately represented as categorical data. This would involve re-coding the data into a limited number of fixed levels (or factors). Some variables to consider for this treatment are authorLang (e.g., levels: English, French, Spanish, German, other), authorLocation (e.g., places grouped by continent), authorPicture (e.g., binomial: field or not) and tweetMethod (e.g., levels: app, website, other). A close examination of each variable is necessary to determine the suitable levels.</p>
<p>There is some duplicate information contained in the dataset. One variable -- "time" -- was dropped in line 31 for this reason. Similarly, authorID and authorID_original are redundant with authorScreen and authorScreen_original. The "ID" variables provide the account numbers associated with the tweets, whereas the "screen" variables provide the screennames.</p>
<p>Given the above information, and to facilitate preparation of the descriptive statistics, I will further process the dataset as follows:
    <ol>
        <li>Drop authorID and authorID_original.</li>
        <li>Drop tweetGeo.</li>
        <li>Convert authorFriend from float64 to int64.</li>
        <li>Convert tweetID and tweetID_original from numerical to object.</li>
        <li>Create new categorical variables and recode into them data from authorLang, authorLocation, authorPicture and tweetMethod (keeping original variables).</li>
        <li>Parse tweetTimeDate into a pandas Timestamp object.</li></ol></p>
        
<p>Note: While I do not expect the tweetURL variable to factor directly into our analysis, I propose keeping it in the dataset for reference purposes. </p>

<h3>Additional data processing prior to analysis</h3>

In [36]:
#1 and 2. Drop authorID, authorID_original and tweetGeo
df = df.drop(['authorID','authorID_original','tweetGeo'], axis=1)
df[1:2]

authorFollower  authorFriend authorLang  authorLocation  \
1  940             225.0         en         top your momma   

                                                                 authorPicture  \
1  http://pbs.twimg.com/profile_images/1062099415421411333/EZCmOCeu_normal.jpg   

    authorScreen authorScreen_original       tweetID  tweetID_original  \
1  BlueelvisEric  NaN                   1.102720e+18 NaN                 

                                                                            tweetMethod  \
1  <a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>   

                    tweetTimeDate  \
1  Mon Mar 04 23:59:40 +0000 2019   

                                                        tweetURL  
1  http://twitter.com/BlueelvisEric/statuses/1102720294035353606

In [ ]:
#3. Convert authorFriend from float64 to int64.
df['authorFriend'] = df['authorFriend'].fillna(0)#Replace NaN with zeros (to represent zero friends)
df['authorFriend'] = df['authorFriend'].astype("int")#Convert to int64
df.dtypes['authorFriend']#Check

In [67]:
#4. Convert tweetID and tweetID_original from numerical to object.
cols = ["tweetID","tweetID_original"]
df[cols] = df[cols].astype("str")
df.dtypes[cols]

tweetID             object
tweetID_original    object
dtype: object

<h3>Notes and code for later use</h3>

In [6]:
users = df['from_user'].tolist()
users = set(users)
print(len(users))

88304


In [16]:
df['user_lang'].value_counts()

en         106942
fr         4523  
de         3016  
es         2488  
en-gb      2107  
nl         735   
it         685   
ja         656   
sv         470   
pt         334   
en-GB      292   
pl         221   
ar         134   
ca         117   
da         112   
fi         83    
tr         82    
ru         77    
el         71    
ko         57    
id         46    
uk         36    
cs         34    
sr         24    
hu         21    
no         20    
zh-cn      18    
he         15    
zh-CN      14    
lv         10    
ro         9     
hr         8     
zh-tw      8     
fa         7     
xx-lc      7     
bg         7     
th         6     
en-AU      5     
ga         4     
vi         4     
gl         4     
zh-TW      2     
nb         2     
eu         2     
hi         2     
sk         2     
fil        1     
pt-PT      1     
en-US      1     
zh-Hans    1     
msa        1     
Name: user_lang, dtype: int64

In [17]:
from collections import Counter

In [19]:
langs = df['user_lang'].tolist()
Counter(langs) #counter object

Counter({'en': 106942,
         'ja': 656,
         'fi': 83,
         'de': 3016,
         'en-gb': 2107,
         'fr': 4523,
         'es': 2488,
         'nl': 735,
         'en-GB': 292,
         'id': 46,
         'pl': 221,
         'it': 685,
         'ar': 134,
         'ko': 57,
         'zh-TW': 2,
         'ca': 117,
         'tr': 82,
         'el': 71,
         'pt': 334,
         'hu': 21,
         'sv': 470,
         'uk': 36,
         'ru': 77,
         'no': 20,
         'fa': 7,
         'da': 112,
         'zh-CN': 14,
         'xx-lc': 7,
         'ro': 9,
         'pt-PT': 1,
         'eu': 2,
         'bg': 7,
         'cs': 34,
         'hr': 8,
         'sr': 24,
         'he': 15,
         'vi': 4,
         'zh-cn': 18,
         'gl': 4,
         'zh-tw': 8,
         'en-AU': 5,
         'lv': 10,
         'th': 6,
         'nb': 2,
         'hi': 2,
         'en-US': 1,
         'sk': 2,
         'msa': 1,
         'fil': 1,
         'zh-Hans': 1,
         'g

In [22]:
df2 = df[['id_str', 'from_user']]
df2.head()

id_str       from_user
0  1.102720e+18  darleneclarke5
1  1.102720e+18  BlueelvisEric 
2  1.102720e+18  Potus16Abe    
3  1.102720e+18  soozbo        
4  1.102720e+18  jp_dutch

In [30]:
L1 = []
for x in df2.values:
    tup = tuple(x)
    L1.append(tup)
print(L1[1:10])

[(1.10272e+18, 'BlueelvisEric'), (1.10272e+18, 'Potus16Abe'), (1.10272e+18, 'soozbo'), (1.10272e+18, 'jp_dutch'), (1.10272e+18, 'RJ_Kendall'), (1.10272e+18, 'gabrieleverett'), (1.10272e+18, 'lT1g0n8T2'), (1.10272e+18, 'suse_______'), (1.10272e+18, 'ffd365')]


In [32]:
L2 = [tuple(x) for x in df2.values]
print(L2[1:10])

[(1.10272e+18, 'BlueelvisEric'), (1.10272e+18, 'Potus16Abe'), (1.10272e+18, 'soozbo'), (1.10272e+18, 'jp_dutch'), (1.10272e+18, 'RJ_Kendall'), (1.10272e+18, 'gabrieleverett'), (1.10272e+18, 'lT1g0n8T2'), (1.10272e+18, 'suse_______'), (1.10272e+18, 'ffd365')]


In [34]:
L1 == L2

True

In [43]:
import dateutil

In [53]:
data['created_at_new'] = data['created_at'].apply(dateutil.parser.parse, dayfirst=True)
data['created_at_new']

id_str
1.102720e+18   2019-03-04 23:59:47+00:00
1.102720e+18   2019-03-04 23:59:40+00:00
1.102720e+18   2019-03-04 23:59:38+00:00
1.102720e+18   2019-03-04 23:59:34+00:00
1.102720e+18   2019-03-04 23:59:32+00:00
1.102720e+18   2019-03-04 23:59:28+00:00
1.102720e+18   2019-03-04 23:59:26+00:00
1.102720e+18   2019-03-04 23:59:19+00:00
1.102720e+18   2019-03-04 23:59:18+00:00
1.102720e+18   2019-03-04 23:59:11+00:00
1.102720e+18   2019-03-04 23:59:09+00:00
1.102720e+18   2019-03-04 23:59:04+00:00
1.102720e+18   2019-03-04 23:58:58+00:00
1.102720e+18   2019-03-04 23:58:52+00:00
1.102720e+18   2019-03-04 23:58:49+00:00
1.102720e+18   2019-03-04 23:58:47+00:00
1.102720e+18   2019-03-04 23:58:46+00:00
1.102720e+18   2019-03-04 23:58:36+00:00
1.102720e+18   2019-03-04 23:58:36+00:00
1.102720e+18   2019-03-04 23:58:34+00:00
1.102720e+18   2019-03-04 23:58:33+00:00
1.102720e+18   2019-03-04 23:58:33+00:00
1.102720e+18   2019-03-04 23:58:32+00:00
1.102720e+18   2019-03-04 23:58:31+00:00
1.102720e

In [54]:
data['time_new'] = data['time'].apply(dateutil.parser.parse, dayfirst=True)
data['time_new']

id_str
1.102720e+18   2019-03-04 23:59:47
1.102720e+18   2019-03-04 23:59:40
1.102720e+18   2019-03-04 23:59:38
1.102720e+18   2019-03-04 23:59:34
1.102720e+18   2019-03-04 23:59:32
1.102720e+18   2019-03-04 23:59:28
1.102720e+18   2019-03-04 23:59:26
1.102720e+18   2019-03-04 23:59:19
1.102720e+18   2019-03-04 23:59:18
1.102720e+18   2019-03-04 23:59:11
1.102720e+18   2019-03-04 23:59:09
1.102720e+18   2019-03-04 23:59:04
1.102720e+18   2019-03-04 23:58:58
1.102720e+18   2019-03-04 23:58:52
1.102720e+18   2019-03-04 23:58:49
1.102720e+18   2019-03-04 23:58:47
1.102720e+18   2019-03-04 23:58:46
1.102720e+18   2019-03-04 23:58:36
1.102720e+18   2019-03-04 23:58:36
1.102720e+18   2019-03-04 23:58:34
1.102720e+18   2019-03-04 23:58:33
1.102720e+18   2019-03-04 23:58:33
1.102720e+18   2019-03-04 23:58:32
1.102720e+18   2019-03-04 23:58:31
1.102720e+18   2019-03-04 23:58:30
1.102720e+18   2019-03-04 23:58:27
1.102720e+18   2019-03-04 23:58:27
1.102720e+18   2019-03-04 23:58:20
1.102720e+18 

In [55]:
 data['from_user'].describe()

count     123524         
unique    88304          
top       ac_migrant_er34
freq      132            
Name: from_user, dtype: object

<h3>References</h3>

<p>Moffitt, C. (2018, March 26) Overview of Pandas Data Types. <i>Practical Business Python</i>. Retrieved March 20, 2019 from https://pbpython.com/pandas_dtypes.html</p>
<p>Lynn, S. (n.d.) Summarising, Aggregating, and Grouping data in Python Pandas. Retrieved March 20, 2019 from https://www.shanelynn.ie/summarising-aggregation-and-grouping-data-in-python-pandas/ </p>
<p>Handling Missing Data (n.d.) <i>Python Data Science Handbook</i>. Retrieved March 20, 2019 from https://jakevdp.github.io/PythonDataScienceHandbook/03.04-missing-values.html</p>
<p>Working with Time Series (n.d.) <i>Python Data Science Handbook</i>. Retrieved March 20, 2019 from https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html</p>